util/metrics.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const metricsNamespace = "epr"
+
+var (
+	httpInFlightRequests = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: metricsNamespace,
+		Name:      "in_flight_requests",
+		Help:      "A gauge of requests currently being served by the http server.",
+	})
+
+	httpRequestsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_requests_total",
+			Help:      "A counter for requests to the http server.",
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpRequestDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_duration_seconds",
+			Help:      "A histogram of latencies for requests to the http server.",
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpRequestSizeBytes = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_size_bytes",
+			Help:      "A histogram of sizes of requests to the http server.",
+			Buckets: []float64{
+				16,
+				32,
+				64,
+				128,
+				256,
+				512,
+				1024,             /* 1KiB */
+				64 * 1024,        /* 64KiB */
+				256 * 1024,       /* 256KiB */
+				1024 * 1024,      /* 1MiB */
+				64 * 1024 * 1024, /* 64MiB */
+			},
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpResponseSizeBytes = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_response_size_bytes",
+			Help:      "A histogram of response sizes for requests to the http server.",
+			Buckets: []float64{
+				10,
+				64,
+				256,
+				512,
+				1024,              /* 1KiB */
+				64 * 1024,         /* 64KiB */
+				256 * 1024,        /* 256KiB */
+				512 * 1024,        /* 512KiB */
+				1024 * 1024,       /* 1MiB */
+				64 * 1024 * 1024,  /* 64MiB */
+				512 * 1024 * 1024, /* 512MiB */
+			},
+		},
+		[]string{"code", "method", "path"},
+	)
+)
+
+// MetricsMiddleware is a middleware used to metric every request received
+func MeticsMiddleware() mux.MiddlewareFunc {
+	// Rergister all metrics
+	prometheus.MustRegister(httpInFlightRequests)
+	prometheus.MustRegister(httpRequestsTotal)
+	prometheus.MustRegister(httpRequestDurationSeconds)
+	prometheus.MustRegister(httpRequestSizeBytes)
+	prometheus.MustRegister(httpResponseSizeBytes)
+
+	return func(next http.Handler) http.Handler {
+		handler := next
+
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			route := mux.CurrentRoute(req)
+			path, _ := route.GetPathTemplate()
+			if req.RequestURI == "/metrics" {
+				next.ServeHTTP(w, req)
+				return
+			}
+			labels := prometheus.Labels{"path": path}
+			handler = promhttp.InstrumentHandlerCounter(httpRequestsTotal.MustCurryWith(labels), handler)
+			handler = promhttp.InstrumentHandlerDuration(httpRequestDurationSeconds.MustCurryWith(labels), handler)
```

Comment:
APM also collects this kind of duration information. Though I think it is ok to have this also as a metric.